### ISSUE TITLE
gpus_ReturnNotPermittedKillClient crash on iOS

### ISSUE THREAD

**Issue by seeuli** [flutter/flutter#62403]
[demo]https://github.com/seeuli/my_app
Flutter Version: 1.12.13、1.9.1
iOS Device：iPhone6

1. Run `flutter packages get`.
2. Open `my_app/example/ios/Runner.xcworkspace`
3. When timer countdown ends, switch application to background, then you will see this crash.

**Expected results:** 
Application crashes on iPhone6.

@jason-simmons @jmagman @gspencergoog 

**Actual results:** 
<img width="1440" alt="snapshot" src="https://user-images.githubusercontent.com/16219433/88659437-9f8cc300-d107-11ea-9478-47b4a2a467ae.png">

## flutter doctor

```
[✓] Flutter (Channel unknown, v1.12.13+hotfix.9, on Mac OS X 10.15.6 19G73, locale zh-Hans-CN)
    • Flutter version 1.12.13+hotfix.9 at /Users/xx/flutter
    • Framework revision f139b11009 (4 months ago), 2020-03-30 13:57:30 -0700
    • Engine revision af51afceb8
    • Dart version 2.7.2
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/xx/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/xx/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 33.4.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.47.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (1 available)
    • iPhone • xxx • ios • iOS 12.4

• No issues found!
```

</details>

**Comment by cwlp8024** [comment]
sixsixsix

**Comment by keshiim** [comment]
I got the same issues, please fix it

**Comment by TahaTesser** [comment]
Hi @seeuli 
No crash on XR running iOS 13

<details> 
<summary>logs</summary> 

``` 
2020-07-28 21:27:55.328212+0530 Runner[7423:2728768] [VERBOSE-1:callback_cache.cc(132)] Could not parse callback cache, aborting restore
didFinishLaunchingWithOptions
2020-07-28 21:27:55.410095+0530 Runner[7423:2728768] -[ViewController timerInvoke], 1
applicationDidBecomeActive
2020-07-28 21:27:56.411409+0530 Runner[7423:2728768] -[ViewController timerInvoke], 2
2020-07-28 21:27:57.411047+0530 Runner[7423:2728768] -[ViewController timerInvoke], 3
2020-07-28 21:27:58.411382+0530 Runner[7423:2728768] -[ViewController timerInvoke], 4
2020-07-28 21:27:59.411406+0530 Runner[7423:2728768] -[ViewController timerInvoke], 5
2020-07-28 21:28:00.566087+0530 Runner[7423:2728768] Metal API Validation Enabled
2020-07-28 21:28:00.800668+0530 Runner[7423:2728768] -[MyFlutterViewController viewDidLoad]
2020-07-28 21:28:00.830143+0530 Runner[7423:2729057] flutter: Observatory listening on http://127.0.0.1:63053/EAvrIyLqV8Y=/
2020-07-28 21:28:01.532625+0530 Runner[7423:2728768] -[MyFlutterViewController viewDidAppear:]
applicationWillResignActive
applicationDidEnterBackground
2020-07-28 21:28:06.354603+0530 Runner[7423:2729045] flutter: Future.delayed
applicationDidBecomeActive
2020-07-28 21:28:07.306542+0530 Runner[7423:2729045] flutter: _TestPageState.initState
2020-07-28 21:28:07.306699+0530 Runner[7423:2729045] flutter: _TestPageState.didChangeDependencies
2020-07-28 21:28:07.307399+0530 Runner[7423:2729045] flutter: _TestPageState.build
applicationWillResignActive
applicationDidEnterBackground
``` 

</details> 


<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.6 19G73, locale en-GB)
    • Flutter version 1.17.5 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 8af6b2f038 (4 weeks ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.1
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (2 available)
    • SM M305F      • 32003c30dc19668f          • android-arm64 • Android 10 (API 29)
    • Taha’s iPhone • 00008020-001059882212002E • ios           • iOS 13.6

• No issues found!
```

</details> 

Looks like you're running 4 months old stable, can you confirm the issue on latest flutter
Can you upgrade to latest `stable` and try again
`flutter channel stable`
`flutter upgrade --force`
`flutter doctor -v`

Thank you

**Comment by seeuli** [comment]
@TahaTesser Please use **_iPhone6_** test, Thank you.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.6 19G73, locale zh-Hans-CN)
    • Flutter version 1.17.5 at /Users/xx/flutter
    • Framework revision 8af6b2f038 (4 weeks ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/xx/Library/Android/sdk
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/xx/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 33.4.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.47.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (1 available)
    • iPhone • xx • ios • iOS 12.4

• No issues found!
```
</details>

<details>
<summary>When using Matel on 1.17.5, on crash but there's something wrong. The detail log</summary>
 
```
2020-07-29 10:24:14.217475+0800 Runner[4228:315634] [VERBOSE-1:callback_cache.cc(132)] Could not parse callback cache, aborting restore
didFinishLaunchingWithOptions
2020-07-29 10:24:14.249559+0800 Runner[4228:315634] -[ViewController timerInvoke], 1
applicationDidBecomeActive
2020-07-29 10:24:15.250781+0800 Runner[4228:315634] -[ViewController timerInvoke], 2
2020-07-29 10:24:16.250796+0800 Runner[4228:315634] -[ViewController timerInvoke], 3
2020-07-29 10:24:17.250762+0800 Runner[4228:315634] -[ViewController timerInvoke], 4
2020-07-29 10:24:18.250747+0800 Runner[4228:315634] -[ViewController timerInvoke], 5
applicationWillResignActive
2020-07-29 10:24:19.267850+0800 Runner[4228:315634] Metal API Validation Enabled
2020-07-29 10:24:19.567793+0800 Runner[4228:315634] -[MyFlutterViewController viewDidLoad]
2020-07-29 10:24:19.914014+0800 Runner[4228:315820] flutter: Observatory listening on http://127.0.0.1:63712/89niiK0NCSA=/
applicationDidEnterBackground
2020-07-29 10:24:20.958594+0800 Runner[4228:315634] -[MyFlutterViewController viewDidAppear:]
2020-07-29 10:24:21.080358+0800 Runner[4228:315787] Execution of the command buffer was aborted due to an error during execution. Insufficient Permission (to submit GPU work from background) (IOAF code 6)
....
2020-07-29 10:24:21.089984+0800 Runner[4228:315787] Execution of the command buffer was aborted due to an error during execution. Insufficient Permission (to submit GPU work from background) (IOAF code 6)
Error submitting command buffer: Insufficient Permission (to submit GPU work from background) (IOAF code 6)
2020-07-29 10:24:22.754555+0800 Runner[4228:315634] Could not load IOSurface for time string. Rendering locally instead.
```
</details>

`_TestPageState` snapshot 
<img src="https://user-images.githubusercontent.com/16219433/88752319-38fdb880-d18c-11ea-9403-963f075828da.PNG" alt="TestPage" width="201" height="384"/>


Steps:
1. I keep application in foreground, everything is ok; 
[forground.MOV.zip](https://github.com/flutter/flutter/files/4992183/forground.MOV.zip)
2. Switch to background when timer countdown ends, until `_TestPageState.build` excute, but the `_TestPageState` is empty;
[background.MOV.zip](https://github.com/flutter/flutter/files/4992239/background.MOV.zip)
3. Switch to background when timer countdown ends, when `Future.delayed` output switch to foreground, and then switch to background, but the `_TestPageState` is also empty;
[switch.MOV.zip](https://github.com/flutter/flutter/files/4992267/switch.MOV.zip)

so, the latest version does not seem to completely solve the problem of froeground and background switching. This problem is currently top1 crash online.
<details>
<summary>crash log</summary>

```
0   libGPUSupportMercury.dylib      gpus_ReturnNotPermittedKillClient + 12
1   AGXGLDriver                     0x00000001a1585ed8 + 468
2   libGPUSupportMercury.dylib      gpusSubmitDataBuffers + 176
3   AGXGLDriver                     0x00000001a1587404 + 292
4   GLEngine                        gliPresentViewES_Exec + 184
5   OpenGLES                        -[EAGLContext presentRenderbuffer:] + 80
6   Flutter                         flutter::IOSSurfaceGL::GLContextPresent() (ios_gl_render_target.mm:0)
7   Flutter                         std::__1::__function::__func<flutter::GPUSurfaceGL::AcquireFrame(SkISize const&)::$_1, std::__1::allocator<flutter::GPUSurfaceGL::AcquireFrame(SkISize const&)::$_1>, bool (flutter::SurfaceFrame const&, SkCanvas*)>::operator()(flutter::SurfaceFrame const&, SkCanvas*&&) (gpu_surface_gl.cc:295)
8   Flutter                         flutter::SurfaceFrame::Submit() (surface.cc:30)
9   Flutter                         flutter::Rasterizer::DrawToSurface(flutter::LayerTree&) (rasterizer.cc:261)
10  Flutter                         std::__1::__function::__func<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0, std::__1::allocator<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0>, void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) (rasterizer.cc:166)
11  Flutter                         flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >) (memory:2635)
12  Flutter                         std::__1::__function::__func<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_28, std::__1::allocator<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_28>, void ()>::operator()() (ref_ptr.h:104)
13  Flutter                         fml::MessageLoopImpl::FlushTasks(fml::FlushType) (ref_ptr.h:116)
14  Flutter                         fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) (message_loop_darwin.mm:76)
15  CoreFoundation                  _CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 28
16  CoreFoundation                  _CFRunLoopDoTimer + 864
17  CoreFoundation                  _CFRunLoopDoTimers + 248
18  CoreFoundation                  _CFRunLoopRun + 1844
19  CoreFoundation                  CFRunLoopRunSpecific + 436
20  GraphicsServices                GSEventRunModal + 104
21  UIKitCore                       UIApplicationMain + 212
22  xxx                           main (main.m:36)
```
</details>

**Comment by TahaTesser** [comment]
I’m unable to reproduce the issue to pinpoint the cause because of specific device unavailability . The issue needs further investigation.

**Comment by seeuli** [comment]
our crash is also on iPhone6, iPhone6p, iPhone5, iPhone5s, try other devices please.

**Comment by PingShen** [comment]
I also meet this issue ,pls fix it as soon as possible.!!!!!!!

**Comment by liangchaojie** [comment]
please fix it ,thanks very much

**Comment by iskakaushik** [comment]
Based on the crash logs it looks the `rasterizer` tried to submit a frame when the app was backgrounded which seems to cause the crash. Looking at https://developer.apple.com/library/archive/qa/qa1766/_index.html, that theory seems to fit. My understanding is that the rasterizer doesn't seem to respect the app being backgrounded.

bef5ac613234ca366c3d73e74cd0adc90a7ec074 should've fixed this issue. @seeuli , do you mind [switching to channel master](https://flutter.dev/docs/development/tools/sdk/upgrading#switching-flutter-channels) and see if you still see the issue?

**Comment by iskakaushik** [comment]
For others seeing this issue, please +1 it rather than adding comments like "i'm seeing this as well". It will help us prioritize.

**Comment by seeuli** [comment]
> When I switch to the master channel, do I use flutter version 1.17? Please provide an address for 'bef5ac613234ca366c3d73e74cd0adc90a7ec074' commit, I will merge this commit based on 1.12 or 1.9

**Comment by iskakaushik** [comment]
Can you try `1.17.5` and let me know if it still fails?

**Comment by seeuli** [comment]
> Can you try `1.17.5` and let me know if it still fails?

I run my demo on 1.17.5 already, please see detail

**Comment by seeuli** [comment]
> Based on the crash logs it looks the `rasterizer` tried to submit a frame when the app was backgrounded which seems to cause the crash. Looking at https://developer.apple.com/library/archive/qa/qa1766/_index.html, that theory seems to fit. My understanding is that the rasterizer doesn't seem to respect the app being backgrounded.
> 
> bef5ac613234ca366c3d73e74cd0adc90a7ec074 should've fixed this issue. @seeuli , do you mind [switching to channel master](https://flutter.dev/docs/development/tools/sdk/upgrading#switching-flutter-channels) and see if you still see the issue?

I tried master channel, everything is ok. Can you release a new stable version? @iskakaushik Thank you very much.

**Comment by seeuli** [comment]
> > When I switch to the master channel, do I use flutter version 1.17? Please provide an address for 'bef5ac613234ca366c3d73e74cd0adc90a7ec074' commit, I will merge this commit based on 1.12 or 1.9

@iskakaushik Is this feasible？

**Comment by jmagman** [comment]
That engine commit from https://github.com/flutter/engine/pull/18563 is available on 1.19.0-2.0.pre, which is available in the [master, dev, and beta channels](https://flutter.dev/docs/development/tools/sdk/releases?tab=macos).
You can read about how we decide to hotfix stable here https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#will-a-particular-bug-fix-be-provided-in-a-hotfix-release.
Closing, thanks for verifying!

**Comment by seeuli** [comment]
> That engine commit from [flutter/engine#18563](https://github.com/flutter/engine/pull/18563) is available on 1.19.0-2.0.pre, which is available in the [master, dev, and beta channels](https://flutter.dev/docs/development/tools/sdk/releases?tab=macos).
> You can read about how we decide to hotfix stable here https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#will-a-particular-bug-fix-be-provided-in-a-hotfix-release.
> Closing, thanks for verifying!


* [flutter/engine#18563](https://github.com/flutter/engine/pull/18563)
* [bef5ac6](https://github.com/flutter/engine/commit/bef5ac613234ca366c3d73e74cd0adc90a7ec074)

The two commits are different, which one should I use？ @jmagman

**Comment by jmagman** [comment]
@seeuli https://github.com/flutter/engine/pull/18563 is the pull request, https://github.com/flutter/engine/commit/bef5ac613234ca366c3d73e74cd0adc90a7ec074 is the commit where that pull request was merged.
The commit is in the underlying rendering engine repository, not this repository.  If you want to cherry-pick a commit, you're signing up for compiling your own version of the engine, which is quite an ordeal.  https://github.com/flutter/flutter/wiki/Compiling-the-engine
And I didn't check but I don't know if that commit merges cleaning onto 1.9.1, that release was branched almost a year ago.
